File: platforms/flutter/colyseus_flutter/lib/src/predict.dart
import 'dart:ffi';

import 'package:ffi/ffi.dart';
import 'package:meta/meta.dart';

import 'bindings/colyseus_core.dart';
import 'colyseus.dart';
import 'events.dart';
import 'input_handle.dart';
import 'interned_names.dart';
import 'reconciler.dart';
import 'room.dart';
import 'schema.dart';
import 'schema_view.dart';
import 'sim_reconciler.dart';
import 'spawns.dart';

/// How a tracked field follows the server stream.
///
/// Values mirror the C `colyseus_predict_mode_t`. They are spelled out rather
/// than read off the generated enum because Dart enum initializers must be
/// compile-time constants; [_assertModesMatchNative] pins them to it.
enum PredictMode {
  /// Render slightly in the past and interpolate between real samples.
  /// Always accurate, never ahead — the default for entities you don't
  /// control.
  lerp(0),

  /// Continue the last known velocity past the newest sample. Hides latency
  /// at the cost of overshooting when the entity changes direction.
  extrapolate(1),

  /// Spring toward the newest sample. No added delay and no overshoot, but
  /// it lags during sustained motion.
  damped(2),

  /// Forward-simulate the last snapshot with a step function shared with the
  /// server. Exact while the entity follows the shared rules.
  reckon(3),

  /// No smoothing — read the last decoded value.
  raw(4);

  final int value;
  const PredictMode(this.value);
}

/// Fails loudly in debug builds if the generated enum ever renumbers.
bool _assertModesMatchNative() {
  assert(
    PredictMode.lerp.value == colyseus_predict_mode_t.COLYSEUS_PREDICT_LERP.value &&
        PredictMode.extrapolate.value ==
            colyseus_predict_mode_t.COLYSEUS_PREDICT_EXTRAPOLATE.value &&
        PredictMode.damped.value ==
            colyseus_predict_mode_t.COLYSEUS_PREDICT_DAMPED.value &&
        PredictMode.reckon.value ==
            colyseus_predict_mode_t.COLYSEUS_PREDICT_RECKON.value &&
        PredictMode.raw.value == colyseus_predict_mode_t.COLYSEUS_PREDICT_RAW.value,
    'PredictMode has drifted from colyseus_predict_mode_t',
  );
  return true;
}

/// Per-field smoothing settings. Zero values take the reference defaults.
class FieldOptions {
  final PredictMode mode;

  /// Interpolation delay in milliseconds for [PredictMode.lerp] (default 100).
  ///
  /// Must match what the app actually displays: with lag compensation the
  /// server rewinds to this instant, so a mismatch makes it rewind to a frame
  /// the player never saw.
  final double delay;

  /// Spring constant (1/s) for damped and extrapolate (default 15).
  final double damping;

  /// Cap on how far extrapolation may run past the newest sample, in
  /// milliseconds (default 200).
  final double maxExtrapolate;

  /// Snap samples to an arrival grid of this many milliseconds; 0 is off.
  final double tickInterval;

  /// Distance beyond which a change is treated as a teleport and applied
  /// instantly instead of smoothed; 0 is off.
  final double snap;

  /// Treat the field as radians and take the shortest arc between samples.
  final bool angle;

  const FieldOptions({
    this.mode = PredictMode.lerp,
    this.delay = 0,
    this.damping = 0,
    this.maxExtrapolate = 0,
    this.tickInterval = 0,
    this.snap = 0,
    this.angle = false,
  });

  /// Fills a native options struct.
  void _write(Pointer<colyseus_predict_field_options_t> out) {
    assert(_assertModesMatchNative());
    out.ref.modeAsInt = mode.value;
    out.ref.delay = delay;
    out.ref.damping = damping;
    out.ref.max_extrapolate = maxExtrapolate;
    out.ref.tick_interval = tickInterval;
    out.ref.snap = snap;
    out.ref.angle = angle;
  }
}

/// Advances a dead-reckoned entity in place.
///
/// [state] is a scratch copy of the latest snapshot, [dt] the step in seconds
/// and [elapsedMs] the absolute server time at the end of the step. Must
/// integrate the same motion the server does, or the entity will drift.
typedef ReckonStep = void Function(
  SchemaView state,
  double dt,
  double elapsedMs,
);

/// Smoothing and prediction for a room's entities.
///
/// The problem it solves: state arrives at the patch rate (20 Hz for the
/// prediction playground) but rendering runs at 60+ Hz. Reading decoded state
/// directly gives visible stepping, and for the entity the player controls it
/// also gives a full round-trip of input lag.
///
/// Two halves:
/// - Entities you don't control are *smoothed* — [attachAll] and [value].
/// - The entity you do control is *predicted and reconciled* — [reconciler].
///
/// Call [tick] once per frame; it drives everything, and returns how many
/// inputs are due:
///
/// ```dart
/// final predict = Predict.of(room);
/// predict.attachAll('players', config: {'x': PredictMode.damped,
///                                       'y': PredictMode.damped},
///                   exceptKey: room.sessionId);
///
/// void frame(double now) {
///   final steps = predict.tick(now);
///   for (var i = 0; i < steps; i++) { input.data['moveX'] = ...; input.send(); }
///   for (final p in players.values) {
///     draw(predict.value(p, 'x'), predict.value(p, 'y'));
///   }
/// }
/// ```
class Predict {
  final Pointer<colyseus_predict> _handle;
  final ColyseusRoom _room;
  final List<NativeCallable> _callables = [];
  final List<Reconciler> _children = [];
  final List<EventChannel> _channels = [];
  final List<Spawns> _spawns = [];
  bool _disposed = false;

  Predict._(this._handle, this._room);

  /// The prediction layer for [room], wired to its clock and input channel.
  ///
  /// Each call builds a new one; hold onto the result and [dispose] it when
  /// the screen goes away. Disposing the room also disposes it, so it can
  /// never outlive the decoder its callbacks live on.
  factory Predict.of(ColyseusRoom room) {
    final native = room.nativeRoom;
    if (native == nullptr) {
      throw StateError('Room is not connected yet — await the join first');
    }
    final predict = Predict._(core.colyseus_predict_for_room(native), room);
    room.registerPredict(predict);
    return predict;
  }

  /// The native handle.
  Pointer<colyseus_predict> get handle => _handle;

  /// Smooths [instance]'s fields per [config].
  ///
  /// Values are a [PredictMode] or a [FieldOptions]. Fields the instance's
  /// schema doesn't declare are ignored, so one config can cover a mixed
  /// collection.
  void attach(SchemaInstance instance, Map<String, Object> config) {
    _withConfig(config, (cfg, count) {
      core.colyseus_predict_attach(
        _handle,
        Pointer<colyseus_schema_t>.fromAddress(instance.handle),
        cfg,
        count,
      );
    });
  }

  /// Smooths every entry of [collection], including ones that arrive later,
  /// and stops when they are removed.
  ///
  /// Pass [exceptKey] your own session id to leave your entity to the
  /// reconciler. With no [config], every numeric field is tracked using
  /// [fallback].
  void attachAll(
    String collection, {
    Map<String, Object>? config,
    String? exceptKey,
    FieldOptions? fallback,
  }) {
    final state = _rootState();
    using((arena) {
      final collectionPtr = collection.toNativeUtf8(allocator: arena);
      final exceptPtr = exceptKey == null
          ? nullptr
          : exceptKey.toNativeUtf8(allocator: arena);
      final fallbackPtr = fallback == null
          ? nullptr
          : (arena<colyseus_predict_field_options_t>()..let(fallback._write));

      if (config == null) {
        core.colyseus_predict_attach_all(_handle, state, collectionPtr.cast(),
            nullptr, 0, exceptPtr.cast(), fallbackPtr);
        return;
      }

      _withConfig(config, (cfg, count) {
        core.colyseus_predict_attach_all(_handle, state, collectionPtr.cast(),
            cfg, count, exceptPtr.cast(), fallbackPtr);
      });
    });
  }

  /// Dead-reckons every entry of [collection] with [step].
  ///
  /// Use when the entities follow rules the client also knows (patrolling
  /// bots, projectiles): forward-simulating the last snapshot beats
  /// interpolating it, because the result is exact rather than delayed.
  ///
  /// [smoothing] of 0 projects raw, which is right for deterministic motion.
  void attachAllReckon(
    String collection,
    List<String> fields,
    ReckonStep step, {
    double smoothing = 0,
    double substepMs = 0,
    double snap = 0,
  }) {
    final state = _rootState();
    final callable = _reckonCallable(step);

    using((arena) {
      final collectionPtr = collection.toNativeUtf8(allocator: arena);
      final fieldPtrs = arena<Pointer<Char>>(fields.length);
      for (var i = 0; i < fields.length; i++) {
        fieldPtrs[i] = fields[i].toNativeUtf8(allocator: arena).cast();
      }

      core.colyseus_predict_attach_all_reckon(
        _handle,
        state,
        collectionPtr.cast(),
        nullptr, // per-entry vtable — required for reflection schemas
        fieldPtrs,
        fields.length,
        callable.nativeFunction,
        smoothing,
        substepMs,
        snap,
        nullptr,
      );
    });
  }

  /// Dead-reckons one instance with [step].
  void attachReckon(
    SchemaInstance instance,
    List<String> fields,
    ReckonStep step, {
    double smoothing = 0,
    double substepMs = 0,
    double snap = 0,
  }) {
    final callable = _reckonCallable(step);

    using((arena) {
      final fieldPtrs = arena<Pointer<Char>>(fields.length);
      for (var i = 0; i < fields.length; i++) {
        fieldPtrs[i] = fields[i].toNativeUtf8(allocator: arena).cast();
      }

      core.colyseus_predict_attach_reckon(
        _handle,
        Pointer<colyseus_schema_t>.fromAddress(instance.handle),
        nullptr,
        fieldPtrs,
        fields.length,
        callable.nativeFunction,
        smoothing,
        substepMs,
        snap,
        nullptr,
      );
    });
  }

  /// Stops smoothing every field of [instance].
  void detach(SchemaInstance instance) {
    core.colyseus_predict_detach(
      _handle,
      Pointer<colyseus_schema_t>.fromAddress(instance.handle),
    );
  }

  /// Advances one frame and returns how many fixed input steps are due.
  ///
  /// Send exactly one input per returned step to stay on the server's
  /// cadence. Returns 0 until a reconciler advertises the step rate, and never
  /// more than 5 — a hitch drops the backlog instead of firing a burst.
  ///
  /// [now] must come from [RoomClock.now], not `DateTime`.
  int tick(double now) => core.colyseus_predict_tick(_handle, now);

  /// The value to draw [field] at, smoothed per its attach config.
  ///
  /// Falls back to the raw decoded value for untracked fields, so it is always
  /// safe to read through.
  double value(SchemaInstance instance, String field) {
    return core.colyseus_predict_value(
      _handle,
      Pointer<colyseus_schema_t>.fromAddress(instance.handle),
      internedName(field),
    );
  }

  /// The raw reckoned value of [field] at server-time [time].
  ///
  /// For game logic rather than drawing: no smoothing offset is applied, so
  /// two clients asking the same question at the same instant get the same
  /// answer. Reckoning into the past clamps to the last snapshot.
  double valueAt(SchemaInstance instance, String field, double time) {
    return core.colyseus_predict_value_at(
      _handle,
      Pointer<colyseus_schema_t>.fromAddress(instance.handle),
      internedName(field),
      time,
    );
  }

  /// Creates a reconciler for the entity this client controls.
  ///
  /// See [Reconciler] for what it does and how [step] must behave.
  Reconciler reconciler(
    SchemaInstance truth, {
    required InputHandle input,
    required ReconcilerStep step,
    List<String>? fields,
    double smoothing = -1,
    double snap = 0,
    double stepMs = 0,
    int subSteps = 0,
    void Function(int ackedSeq)? onReconcile,
  }) {
    final child = Reconciler.create(
      predict: this,
      truth: truth,
      input: input,
      step: step,
      fields: fields,
      smoothing: smoothing,
      snap: snap,
      stepMs: stepMs,
      subSteps: subSteps,
      onReconcile: onReconcile,
    );
    return child;
  }

  /// Creates an optimistic-event channel driven by [tick].
  ///
  /// See [EventChannel] for what it does. [graceTicks] is how far the server
  /// may progress past a sim-born prediction before it auto-rejects.
  EventChannel defineEvent({
    void Function(Object? payload)? onPredict,
    void Function(Object? payload)? onConfirm,
    void Function(Object? payload)? onReject,
    void Function(String key)? onUnpredicted,
    int graceTicks = 0,
    double ttlMs = 0,
    double cooldownMs = 0,
  }) {
    final channel = EventChannel.create(
      clock: _room.clock.handle,
      onPredict: onPredict,
      onConfirm: onConfirm,
      onReject: onReject,
      onUnpredicted: onUnpredicted,
      graceTicks: graceTicks,
      ttlMs: ttlMs,
      cooldownMs: cooldownMs,
    );
    core.colyseus_predict_drive_events(_handle, channel.handle);
    _channels.add(channel);
    return channel;
  }

  /// Creates a predicted-spawn store bound to [collection].
  ///
  /// The collection's additions and removals are routed into the store, and
  /// the store is advanced by [tick] — this is the whole wiring a predicted
  /// collection needs. See [Spawns].
  Spawns spawns(String collection, SpawnsOptions options) {
    final store = Spawns.create(clock: _room.clock.handle, options: options);
    final state = _rootState();

    using((arena) {
      final collectionPtr = collection.toNativeUtf8(allocator: arena);

      Pointer<colyseus_spawns_reckon_t> reckon = nullptr;
      if (options.fields != null && options.reckonStep != null) {
        final fields = options.fields!;
        final names = arena<Pointer<Char>>(fields.length);
        for (var i = 0; i < fields.length; i++) {
          names[i] = fields[i].toNativeUtf8(allocator: arena).cast();
        }

        final callable = _reckonCallable(
          (state, dt, elapsedMs) => options.reckonStep!(state, dt, elapsedMs),
        );

        reckon = arena<colyseus_spawns_reckon_t>();
        reckon.ref.fields = names;
        reckon.ref.field_count = fields.length;
        reckon.ref.step = callable.nativeFunction;
        // Raw projection: a deterministic constant-step projectile rebases
        // exactly, so smoothing would only add lag.
        reckon.ref.smoothing = 0;
      }

      core.colyseus_predict_bind_spawns(
        _handle,
        store.handle,
        state,
        collectionPtr.cast(),
        reckon,
      );
    });

    _spawns.add(store);
    return store;
  }

  /// Releases the prediction layer and everything it drives.
  ///
  /// Safe to call twice, and called for you by [ColyseusRoom.dispose].
  void dispose() {
    if (_disposed) return;
    _disposed = true;

    for (final child in _children.toList()) {
      child.dispose();
    }
    _children.clear();

    // Children deregister themselves from the driver as they go, so they all
    // have to be gone before the Predict itself is freed.
    for (final channel in _channels) {
      channel.dispose();
    }
    _channels.clear();
    for (final store in _spawns) {
      store.dispose();
    }
    _spawns.clear();

    core.colyseus_predict_free(_handle);
    _room.unregisterPredict(this);

    for (final callable in _callables) {
      callable.close();
    }
    _callables.clear();
  }

  @internal
  /// Called by a child that was disposed on its own.
  void forget(Reconciler child) => _children.remove(child);

  @internal
  /// Registers a reconciler built outside [reconciler] for ordered teardown.
  void adoptChild(Reconciler child) {
    if (!_children.contains(child)) _children.add(child);
  }

  /// Creates a composite reconciler over several interacting entities.
  ///
  /// See [createSimReconciler] for when this beats a flat [reconciler].
  Reconciler sim({
    required InputHandle input,
    required SimStep step,
    required SimOptions options,
  }) {
    return createSimReconciler(
      predict: this,
      input: input,
      step: step,
      options: options,
    );
  }

  /// The root state, re-read every time.
  ///
  /// The decoder can free and replace instances, so caching the pointer would
  /// leave a dangling read after a resync.
  Pointer<colyseus_schema_t> _rootState() {
    final state = _room.state;
    if (state == null) {
      throw StateError('Room state has not decoded yet');
    }
    return Pointer<colyseus_schema_t>.fromAddress(state.handle);
  }

  /// Wraps [step] in a native callable and keeps it alive for this Predict.
  ///
  /// The core calls it synchronously from inside `tick()`, which Dart
  /// initiated, so it runs on this isolate's own thread.
  NativeCallable<Void Function(Pointer<colyseus_schema_t>, Double, Double,
      Pointer<Void>)> _reckonCallable(ReckonStep step) {
    // One view per scratch pointer: the core reuses a small set of scratch
    // instances, so this cache stays tiny and skips re-resolving field names.
    final views = <int, SchemaView>{};

    final callable = NativeCallable<
        Void Function(Pointer<colyseus_schema_t>, Double, Double,
            Pointer<Void>)>.isolateLocal((
      Pointer<colyseus_schema_t> state,
      double dt,
      double elapsedMs,
      Pointer<Void> _,
    ) {
      final view = views.putIfAbsent(state.address, () => SchemaView(state.address));
      try {
        step(view, dt, elapsedMs);
      } catch (_) {
        // An exception here would unwind through C and abort; a bad step
        // should leave the entity where it was instead.
      }
    });

    _callables.add(callable);
    return callable;
  }

  /// Marshals a Dart config map into the native array of field configs.
  void _withConfig(
    Map<String, Object> config,
    void Function(Pointer<colyseus_attach_field_t> cfg, int count) body,
  ) {
    using((arena) {
      final entries = config.entries.toList();
      final cfg = arena<colyseus_attach_field_t>(entries.length);

      for (var i = 0; i < entries.length; i++) {
        final value = entries[i].value;
        final options = switch (value) {
          PredictMode mode => FieldOptions(mode: mode),
          FieldOptions opts => opts,
          _ => throw ArgumentError(
              'Field "${entries[i].key}" must map to a PredictMode or '
              'FieldOptions, got ${value.runtimeType}'),
        };

        final optsPtr = arena<colyseus_predict_field_options_t>();
        options._write(optsPtr);
        cfg[i].field = entries[i].key.toNativeUtf8(allocator: arena).cast();
        cfg[i].opts = optsPtr;
      }

      body(cfg, entries.length);
    });
  }
}

extension<T extends NativeType> on Pointer<T> {
  /// Applies [body] to this pointer and returns it, for inline struct fills.
  Pointer<T> let(void Function(Pointer<T>) body) {
    body(this);
    return this;
  }
}

File: platforms/swift/Sources/Colyseus/Predict.swift
import CColyseus
import Foundation

public extension Colyseus {
    /// How one field is smoothed between server snapshots.
    enum PredictMode: Sendable {
        /// Interpolate between the two snapshots either side of render time.
        /// Accurate, and a render delay behind.
        case lerp
        /// Project past the newest snapshot along its velocity. No delay, and
        /// wrong the moment the entity turns.
        case extrapolate
        /// Chase the newest value with an exponential decay. Never overshoots,
        /// always trailing.
        case damped
        /// Forward-simulate the last snapshot through a step function shared
        /// with the server.
        case reckon
        /// Draw the decoded value as it arrives, steps and all.
        case raw

        var asC: colyseus_predict_mode_t {
            switch self {
            case .lerp: return COLYSEUS_PREDICT_LERP
            case .extrapolate: return COLYSEUS_PREDICT_EXTRAPOLATE
            case .damped: return COLYSEUS_PREDICT_DAMPED
            case .reckon: return COLYSEUS_PREDICT_RECKON
            case .raw: return COLYSEUS_PREDICT_RAW
            }
        }
    }

    /// Per-field smoothing settings. Anything left alone takes the reference
    /// default: 100 ms render delay, 50 ms smoothing on damped and
    /// extrapolate, 200 ms extrapolation cap.
    struct PredictOptions: Sendable {
        public var mode: PredictMode
        /// Render-time lag for `lerp`, in milliseconds.
        public var delay: Double = 0
        /// Output-smoothing time constant, in milliseconds — roughly the extra
        /// display lag it buys. Pass a negative number for an explicit zero.
        public var smoothMs: Double = 0
        /// How far past the newest snapshot `extrapolate` may reach.
        public var maxExtrapolate: Double = 0
        /// Snap arrival timestamps to this grid, in milliseconds. 0 is off.
        public var snap: Double = 0
        /// Treat the field as radians and take the shortest arc between samples.
        public var isAngle: Bool = false

        public init(
            mode: PredictMode = .lerp,
            delay: Double = 0,
            smoothMs: Double = 0,
            maxExtrapolate: Double = 0,
            snap: Double = 0,
            isAngle: Bool = false
        ) {
            self.mode = mode
            self.delay = delay
            self.smoothMs = smoothMs
            self.maxExtrapolate = maxExtrapolate
            self.snap = snap
            self.isAngle = isAngle
        }

        var asC: colyseus_predict_field_options_t {
            var options = colyseus_predict_field_options_t()
            options.mode = mode.asC
            options.delay = delay
            options.smooth_ms = smoothMs
            options.max_extrapolate = maxExtrapolate
            options.snap = snap
            options.angle = isAngle
            return options
        }
    }

    /// Smoothing and prediction for one room.
    ///
    /// A Predict owns three jobs that all key off the same frame tick: it
    /// smooths entities you do not control, it hosts the reconciler for the
    /// one you do, and it paces your input sends.
    ///
    /// ```swift
    /// let predict = Colyseus.Predict.get(room)
    /// predict.attachAll(state.players, fields: ["x", "y"], except: room.sessionId)
    ///
    /// // once a frame
    /// for _ in 0 ..< predict.tick(room.clock.now) { input.send() }
    /// let x = predict.value(player, "x")
    /// ```
    final class Predict: @unchecked Sendable {
        let raw: OpaquePointer
        /// The room clock, which the children this Predict creates need too.
        let clock: OpaquePointer?

        /// The room this Predict borrows its callbacks layer from.
        ///
        /// Held strongly, because freeing the room frees that layer: a Predict
        /// that outlived its room would unregister from memory that is gone.
        /// Ownership runs one way — reconciler to Predict to room — so
        /// teardown order follows without anyone having to remember it.
        private let room: AnyObject?

        /// Callbacks the C side holds by pointer for this Predict's lifetime
        /// and never hands back: reckoning steps, which have no detach hook.
        private var retained: [AnyObject] = []
        private let lock = NSLock()

        init(raw: OpaquePointer, clock: OpaquePointer?, room: AnyObject?) {
            self.raw = raw
            self.clock = clock
            self.room = room
        }

        deinit { colyseus_predict_free(raw) }

        /// The Predict for this room: its own callbacks layer, wired to the
        /// room clock and the server's fixed step.
        public static func get<State: SchemaRef>(_ room: Room<State>) -> Predict? {
            guard let raw = colyseus_predict_for_room(room.raw) else { return nil }
            return Predict(raw: raw, clock: room.clock.raw, room: room)
        }

        // MARK: - Attaching

        /// Smooth these fields of one instance.
        @discardableResult
        public func attach(_ instance: SchemaRef, fields: [String: PredictOptions]) -> Bool {
            withFieldConfig(fields) { config in
                colyseus_predict_attach(raw, instance.view.instance, config, Int32(config.count)) == 0
            }
        }

        /// Smooth these fields of one instance, all the same way.
        @discardableResult
        public func attach(
            _ instance: SchemaRef,
            fields: [String],
            options: PredictOptions = .init()
        ) -> Bool {
            attach(instance, fields: Dictionary(uniqueKeysWithValues: fields.map { ($0, options) }))
        }

        /// Smooth every entry of a collection — the ones there now and the
        /// ones that arrive later — and stop when they are removed.
        ///
        /// `except` skips one entry by key, which is how you leave your own
        /// player to the reconciler.
        @discardableResult
        public func attachAll<Element: SchemaValue>(
            _ collection: MapSchema<Element>,
            fields: [String: PredictOptions],
            except key: String? = nil
        ) -> Bool {
            withFieldConfig(fields) { config in
                withCStrings([collection.field, key]) { strings in
                    colyseus_predict_attach_all(
                        raw, collection.owner.view.instance, strings[0],
                        config, Int32(config.count), strings[1], nil
                    ) == 0
                }
            }
        }

        @discardableResult
        public func attachAll<Element: SchemaValue>(
            _ collection: MapSchema<Element>,
            fields: [String],
            options: PredictOptions = .init(),
            except key: String? = nil
        ) -> Bool {
            attachAll(
                collection,
                fields: Dictionary(uniqueKeysWithValues: fields.map { ($0, options) }),
                except: key
            )
        }

        /// Stop smoothing one instance.
        public func detach(_ instance: SchemaRef) {
            colyseus_predict_detach(raw, instance.view.instance)
        }

        // MARK: - Dead reckoning

        /// Forward-simulate every entry of a collection through a step shared
        /// with the server, instead of interpolating its snapshots.
        ///
        /// Where it applies this beats interpolation outright: there is no
        /// render delay, because the client computes what the server is about
        /// to send rather than waiting to be told. Where it does not — an
        /// entity whose next move depends on something only the server knows —
        /// it breaks honestly and visibly.
        ///
        /// The step runs on demand and its result is cached for the frame, so
        /// a loop that ticks without reading ``value(_:_:)`` reckons nothing.
        ///
        /// - Parameters:
        ///   - fields: the numeric fields the step reads and writes.
        ///   - smoothMs: how fast a rebase onto a fresh snapshot eases in.
        ///   - substepMs: integration granularity. 0 takes 16 ms.
        ///   - snap: a rebase further than this cuts instead of gliding — for
        ///     teleports, which are not reckoning failures.
        @discardableResult
        public func attachAllReckon<Element: SchemaValue>(
            _ collection: MapSchema<Element>,
            fields: [String],
            smoothMs: Double = 0,
            substepMs: Double = 0,
            snap: Double = 0,
            step: @escaping @Sendable (SchemaView, Double, Double) -> Void
        ) -> Bool {
            let box = ReckonBox(step)
            retain(box)

            return withCStrings(fields) { fieldPointers in
                var borrowed = fieldPointers
                return borrowed.withUnsafeMutableBufferPointer { buffer in
                    collection.field.withCString { collectionPointer in
                        colyseus_predict_attach_all_reckon(
                            raw,
                            collection.owner.view.instance,
                            collectionPointer,
                            // Each entry reckons with its own vtable, which is
                            // the only option for reflection-built schemas.
                            nil,
                            UnsafePointer(buffer.baseAddress),
                            Int32(fields.count),
                            { state, dt, elapsedMs, userdata in
                                guard let state, let box = borrowObject(userdata, as: ReckonBox.self) else { return }
                                box.body(SchemaView(state), dt, elapsedMs)
                            },
                            smoothMs, substepMs, snap,
                            retainedPointer(box)
                        ) == 0
                    }
                }
            }
        }

        /// Dead-reckon one instance.
        @discardableResult
        public func attachReckon(
            _ instance: SchemaRef,
            fields: [String],
            smoothMs: Double = 0,
            substepMs: Double = 0,
            snap: Double = 0,
            step: @escaping @Sendable (SchemaView, Double, Double) -> Void
        ) -> Bool {
            let box = ReckonBox(step)
            retain(box)

            return withCStrings(fields) { fieldPointers in
                var borrowed = fieldPointers
                return borrowed.withUnsafeMutableBufferPointer { buffer in
                    colyseus_predict_attach_reckon(
                        raw,
                        instance.view.instance,
                        instance.view.instance.pointee.__vtable,
                        UnsafePointer(buffer.baseAddress),
                        Int32(fields.count),
                        { state, dt, elapsedMs, userdata in
                            guard let state, let box = borrowObject(userdata, as: ReckonBox.self) else { return }
                            box.body(SchemaView(state), dt, elapsedMs)
                        },
                        smoothMs, substepMs, snap,
                        retainedPointer(box)
                    ) == 0
                }
            }
        }

        // MARK: - The frame

        /// Advance one render frame, and say how many fixed input steps are due.
        ///
        /// Send exactly one input per returned step and the client stays on the
        /// server's cadence. The count is 0 until a reconciler advertises the
        /// step, and is capped at 5 — a hitch drops the backlog rather than
        /// firing a burst.
        @discardableResult
        public func tick(_ now: Double) -> Int {
            Int(colyseus_predict_tick(raw, now))
        }

        /// What to draw for this field: smoothed, predicted, or the raw
        /// decoded value when the field is not tracked.
        ///
        /// This is the one read a render loop should make — it keeps working
        /// as a field moves between raw, smoothed and reconciled.
        public func value(_ instance: SchemaRef, _ field: String) -> Double {
            field.withCString { colyseus_predict_value(raw, instance.view.instance, $0) }
        }

        /// The raw reckoned value at a given server-time instant — for hit
        /// tests and other game logic, where the smoothing offset would lie.
        public func value(_ instance: SchemaRef, _ field: String, at time: Double) -> Double {
            field.withCString { colyseus_predict_value_at(raw, instance.view.instance, $0, time) }
        }

        // MARK: -

        /// Keep an object alive for this Predict's whole life, for the C
        /// callbacks that are never unregistered.
        func retain(_ object: AnyObject) {
            lock.lock()
            retained.append(object)
            lock.unlock()
        }

        /// The C config is an array of (name, options) — the reference's
        /// per-field map with no map literal to write it in.
        private func withFieldConfig<R>(
            _ fields: [String: PredictOptions],
            _ body: ([colyseus_attach_field_t]) -> R
        ) -> R {
            let names = fields.keys.sorted()
            var options = names.map { fields[$0]!.asC }

            return withCStrings(names) { pointers in
                options.withUnsafeMutableBufferPointer { optionsBuffer in
                    let config = names.indices.map { index in
                        colyseus_attach_field_t(
                            field: pointers[index],
                            opts: optionsBuffer.baseAddress! + index
                        )
                    }
                    return body(config)
                }
            }
        }
    }
}

/// The reckoning step is a C function pointer invoked per substep, so the
/// closure lives in an object userdata can reach. There is no detach hook for
/// it, so the Predict keeps it for its own lifetime.
final class ReckonBox: @unchecked Sendable {
    let body: (SchemaView, Double, Double) -> Void

    init(_ body: @escaping (SchemaView, Double, Double) -> Void) {
        self.body = body
    }
}
